src/main/java/org/opensearch/neuralsearch/search/HitsThresholdChecker.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.neuralsearch.search;
+
+import java.util.Locale;
+
+import org.apache.lucene.search.ScoreMode;
+
+/**
+ *  Defines algorithm to allow searches to terminate early
+ */
+public class HitsThresholdChecker {
+    private int hitCount;
+    private final int totalHitsThreshold;
+
+    private HitsThresholdChecker(int totalHitsThreshold) {
+        if (totalHitsThreshold < 0) {
+            throw new IllegalArgumentException(String.format(Locale.ROOT, "totalHitsThreshold must be >= 0, got %d", totalHitsThreshold));
+        }
+        assert totalHitsThreshold != Integer.MAX_VALUE;
```

Comment:
Should this be an assertion error or should we throw an IllegalArgumentException

Reply:
will change to exception, assertion is not a requirement here